Add doc for AWS XKS Proxy with PKCS#11 Provider (#18149)
hashicorp/vault: website/content/docs/enterprise/pkcs11-provider/aws-xks.mdx

description: |-
  AWS KMS External Key Store can use Vault as a key store via the Vault PKCS#11 Provider.
---

# Vault with AWS KMS External Key Store (XKS) via PKCS#11 and XKS Proxy

~> **Note**: AWS [`xks-proxy`](https://github.com/aws-samples/aws-kms-xks-proxy) is not considered production-ready.

Vault's KMIP Secrets Engine can be used as an external key store for the AWS KMS [External Key Store (XKS)](https://aws.amazon.com/blogs/aws/announcing-aws-kms-external-key-store-xks/) protocol using the AWS [`xks-proxy`](https://github.com/aws-samples/aws-kms-xks-proxy) along
with the [Vault PKCS#11 Provider](https://developer.hashicorp.com/vault/docs/enterprise/pkcs11-provider).

## Overview

This is tested as working with Vault 1.11.0 Enterprise (and later) with Advanced Data Protection (KMIP support).

Prerequisites:

* A server capable of running XKS Proxy, which is exposed to the Internet or a VPC endpoint. This can be the same as the Vault server.
* `libvault-pkcs11.so` downloaded from [releases.hashicorp.com](https://releases.hashicorp.com/vault-pkcs11-provider) for your platform and available on the XKS Proxy server.

> * Vault Enterprise with the KMIP Secrets Engine available and with TCP port 5696 accessible to where XKS Proxy will be running.

Reviewer: should we add: with ADP license?

Author: I believe I mention it above.

---
layout: docs
page_title: AWS KMS External Key Store (XKS) - PKCS#11 Provider - Vault Enterprise
description: |-

>   The Vault PKCS#11 Provider can be used with AWS KMS External Key Store.

Reviewer 2: nit: It probably makes more sense to readers if we start with a more zoomed out view:

```suggestion
  AWS KMS External Key Store can use Vault as a key store via the Vault PKCS#11 Provider.
```

---

# Vault with AWS KMS External Key Store (XKS) via PKCS#11 and XKS Proxy

> ~> **Important**: AWS [`xks-proxy`](https://github.com/aws-samples/aws-kms-xks-proxy) is not considered production-ready.

Reviewer 2: I see it's the reference implementation, but I don't see any mention on that page or the blog of not being production-ready. Am I missing it, or is there another page which references the production readiness directly?

Reviewer 2: nit: For consistency with other callouts? I guess the formatting already implies it's important.

```suggestion
~> **Note**: AWS [`xks-proxy`](https://github.com/aws-samples/aws-kms-xks-proxy) is not considered production-ready.
```

Author: I don't know if they have called out the production readiness of it explicitly or not.

Reviewer 2: Should we delete the callout then?

Author: :+1:

## Overview

This is tested as working with Vault 1.11.0 Enterprise (and later) with Advanced Data Protection (KMIP support).

Prerequisites:

* A server capable of running XKS Proxy, which is exposed to the Internet or a VPC endpoint. This can be the same as the Vault server.
* `libvault-pkcs11.so` downloaded from [releases.hashicorp.com](https://releases.hashicorp.com/vault-pkcs11-provider) for your platform and available on the XKS Proxy server.
* Vault Enterprise with the KMIP Secrets Engine available and with TCP port 5696 accessible to where XKS Proxy will be running.

There are 3 parts to this setup:

1. Vault KMIP Secrets Engine standard setup. (There is nothing specific to XKS in this setup.)
1. Vault PKCS#11 setup to tell the PKCS#11 provider (`libvault-pkcs11.so`) how to talk to the Vault KMIP Secrets Engine. (There is nothing specific to XKS in this setup.)
1. XKS Proxy setup.

## Vault Setup

On the Vault server, we need to [setup the KMIP Secrets Engine](/docs/secrets/kmip):

1. Start the [KMIP Secrets Engine](/docs/secrets/kmip) and listener:

    ```sh
    vault secrets enable kmip
    vault write kmip/config listen_addrs=0.0.0.0:5696
    ```

>     ~> **Important**: When configuring KMIP for Oracle, you will probably need to set the

Reviewer 2: Optional nit if you agree with the similar one above

```suggestion
    ~> **Note**: When configuring KMIP for Oracle, you will probably need to set the
```

    `server_hostnames` and `server_ips` [configuration parameters](/api-docs/secret/kmip#parameters),
    otherwise the TLS connection to the KMIP Secrets Engine will fail due to certification validation errors.

>     When configuring Oracle TDE, this error can manifest as the `sqlplus` session silently hanging.

Reviewer 3: Is this oracle info relevant for XKS?

Author: Ah, good catch. I thought I had deleted all of the Oracle stuff from this doc again. :)

## Enable XKS in the AWS UI

1. Go into KMS settings in the AWS UI.
1. Go to "External key stores" in the navigation menu.
1. Click on "Create external key store"
1. Add the relevant settings:
   a. Key store name: the name you want your key store to be referred to in AWS
   a. Proxy URI endpoint: the base URI of the XKS proxy server (e.g., `https://xks.example.com:1234`)
   a. Proxy URI path prefix: added to the end of the base URI of the XKS proxy server when making requests
   a. Proxy credential, Access key ID: the access key you configured in the XKS proxy configuration file
   a. Proxy credential, Secret access key: the secret key you configured in the XKS proxy configuration file
1. Create the external key store
1. Connect the external key store (will take a few minutes)
1. Select the newly created external key store
1. Select "Create a KMS key in this key store"
   a. Enter an external key ID corresponding to a PKCS#11 label that is allowed in the settings TOML. In the example above, this is "abc123".
   a. Enter an alias for this key.
   a. Set up the key administrators and key usage permissions.

> 1. You should not be able to use this Vault-backed key with the KMS API, e.g., `aws kms encrypt`.

Reviewer 3: Should this say `now` instead of `not`? Or maybe there's some other way to use the key?

```suggestion
1. You should now be able to use this Vault-backed key with the KMS API, e.g., `aws kms encrypt`.
```
